pyra/hardware.py

```diff
@@ -0,0 +1,424 @@
+
+    Examples
+    --------
+    >>> update()
+    """
+    global initialized
+
+    current_timestamp = helpers.timestamp()
+
+    if initialized:
+        dash_stats['timestamp'].append(helpers.timestamp())
+
+        dash_stats['relative_time'] = []
+        for x in dash_stats['timestamp']:
+            seconds_ago = current_timestamp - x
+            dash_stats['relative_time'].append(seconds_ago)
+
+        child_processes = proc.children(recursive=False)  # list all children processes
+
+        for child in child_processes:
+            if child not in processes:
+                processes.append(child)
+
+    for p in processes:
+        # set the name
+        proc_name = definitions.Names().name if p.pid == proc_id else p.name()
+
+        # cpu stats per process
+        proc_cpu_percent = None
+        try:
+            proc_cpu_percent = min(100, p.cpu_percent())  # get current value, max of 100
+        except psutil.NoSuchProcess:
+            pass
+        finally:
+            if initialized:
+                try:
+                    dash_stats[f'cpu_{proc_name}']
+                except KeyError:
+                    dash_stats[f'cpu_{proc_name}'] = []
+                finally:  # append the current value to the list
+                    dash_stats[f'cpu_{proc_name}'].append(proc_cpu_percent)
+
+        # memory stats per process
+        proc_memory_percent = None
+        try:
+            proc_memory_percent = min(100, p.memory_percent(memtype='rss'))  # get current value, max of 100
+        except psutil.NoSuchProcess:
+            pass
+        finally:
+            if initialized:
+                try:
+                    dash_stats[f'memory_{proc_name}']
+                except KeyError:
+                    dash_stats[f'memory_{proc_name}'] = []
+                finally:  # append the current value to the list
+                    dash_stats[f'memory_{proc_name}'].append(proc_memory_percent)
+
+    update_cpu()  # todo, need to investigate why this is sometimes lower than the individual process
+    update_gpu()  # todo... AMD GPUs on non Linux... integrated GPUs... GPU stats for processes
+    update_memory()
+    update_network()  # todo... network stats for processes
+
+    for key in dash_stats:
+        dash_stats[key] = dash_stats[key][-history_length:]  # keep the first 2 minutes
+
+    if not initialized:
+        initialized = True
+
+
+def chart_data() -> str:
+    """
+    Get chart data.
+
+    Get the data from the ``dash_stats`` dictionary, formatted for use with ``plotly``.
+
+    Returns
+    -------
+    str
+        A list, formatted as a string, containing dictionaries.
+        Each dictionary is a chart ready to use with ``plotly``.
+
+    See Also
+    --------
+    pyra.webapp.callback_dashboard : A callback called by javascript to get this data.
+
+    Examples
+    --------
+    >>> chart_data()
+    '[{"data": [...], "layout": ..., "config": ..., {"data": ...]'
+    """
```

Comment:
I think that this function should return a dict or a plotly object and the flask endpoint should deal with encoding it as JSON. That would make it easier to reuse this function in another way. Also, the flask endpoint is really short and this is really long so that would redistribute the complexity well.

Reply:
> I think that this function should return a dict or a plotly object and the flask endpoint should deal with encoding it as JSON. That would make it easier to reuse this function in another way. Also, the flask endpoint is really short and this is really long so that would redistribute the complexity well.

It's a list of 4 dictionaries, so don't think it can be a plotly object, but I can handle the json later.